src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/feature/AbstractUmapFeatureComputer.java

```diff
@@ -0,0 +1,239 @@
+		for ( UmapInputDimension< V > inputDimension : inputDimensions )
+			logger.info( "Input dimension: {}", inputDimension );
+		this.inputDimensions = inputDimensions;
+		this.forceComputeAll = new AtomicBoolean( true );
+		long start = System.currentTimeMillis();
+		ReentrantReadWriteLock.ReadLock lock = getLock( graph ).readLock();
+		lock.lock();
+		try
+		{
+			run();
+		}
+		finally
+		{
+			lock.unlock();
+		}
+		logger.info( "Finished computing UmapFeature in {} ms", System.currentTimeMillis() - start );
+		model.getFeatureModel().declareFeature( feature );
+	}
+
+	@Override
+	protected void compute( final V vertex )
+	{
+		if ( isVertexValid( vertex ) )
+		{
+			for ( int i = 0; i < settings.getNumberOfOutputDimensions(); i++ )
+			{
+				DoublePropertyMap< V > umapOutput = feature.getUmapOutputMaps().get( i );
+				umapOutput.set( vertex, umapResult[ index ][ i ] );
+			}
+			index++;
+		}
+	}
+
+	@Override
+	public void createOutput()
+	{
+		if ( feature == null )
+			feature = initFeature( settings.getNumberOfOutputDimensions() );
+		computeUmap();
+	}
+
+	@Override
+	protected void notifyProgress( final int finished, final int total )
+	{
+		statusService.showStatus( finished, total, "Computing UmapFeature" );
+	}
+
+	@Override
+	protected ValueIsSetEvaluator< V > getEvaluator()
+	{
+		return feature;
+	}
+
+	@Override
+	protected void reset()
+	{
+		if ( feature == null )
+			return;
+		feature.getUmapOutputMaps().forEach( DoublePropertyMap::beforeClearPool );
+	}
+
+	private void computeUmap()
+	{
+		index = 0;
+		List< double[] > data = extractAndCacheValidDataRows();
+		double[][] dataMatrix = data.toArray( new double[ 0 ][ 0 ] );
+		if ( dataMatrix.length == 0 )
+			throw new IllegalArgumentException(
+					"No valid data rows found, i.e. in each existing data row there is at least one non-finite value, such Not a Number or Infinity." );
+		if ( settings.isStandardizeFeatures() )
+		{
+			logger.debug( "Standardizing features with {} rows", dataMatrix.length );
+			for ( int i = 0; i < dataMatrix.length; i++ )
+				dataMatrix[ i ] = StandardScaler.standardizeVector( dataMatrix[ i ] );
+			logger.debug( "Finished standardizing features" );
+		}
+		umap.setNumberComponents( settings.getNumberOfOutputDimensions() );
+		umap.setNumberNearestNeighbours( settings.getNumberOfNeighbors() );
+		umap.setMinDist( ( float ) settings.getMinimumDistance() );
+		umap.setThreads( 1 );
+		umap.setSeed( 42 );
+		logger.info( "Fitting umap. Data matrix has {} rows x {} columns.", dataMatrix.length, inputDimensions.size() );
+		umapResult = umap.fitTransform( dataMatrix );
+		logger.info( "Finished fitting umap. Results has {} rows x {} columns.", umapResult.length,
+				umapResult.length > 0 ? umapResult[ 0 ].length : 0 );
+	}
+
+	private List< double[] > extractAndCacheValidDataRows()
+	{
+		List< double[] > data = new ArrayList<>();
+		for ( V vertex : getVertices() )
```

Comment:
You are using a `Collection<V>` interface here. Don't assume that there is a particular iterations order for a `Collection` it's not guarantied.